### ISSUE TITLE
dead_code false positive introduced in rustc 1.78.0-nightly (8ace7ea1f 2024-02-07) 

### ISSUE THREAD

**Issue by tamird** [rust-lang/rust#120770]
```rust
#![deny(warnings)]

#[repr(u8)]
#[derive(Copy, Clone, Debug)]
pub enum RecordField {
    Target = 1,
    Level,
    Module,
    File,
    Line,
    NumArgs,
}

unsafe trait Pod {}

#[repr(transparent)]
struct RecordFieldWrapper(RecordField);

unsafe impl Pod for RecordFieldWrapper {}

fn try_read<T: Pod>(buf: &[u8]) -> T {
    unsafe { std::ptr::read_unaligned(buf.as_ptr() as *const T) }
}

pub fn foo(buf: &[u8]) -> RecordField {
    let RecordFieldWrapper(tag) = try_read(buf);
    tag
}
```

This code compiles on stable (and nightlies before 2024-02-07) but now produces:
```
error: struct `RecordFieldWrapper` is never constructed
  --> aya-log/src/lib.rs:79:8
   |
79 | struct RecordFieldWrapper(RecordField);
   |        ^^^^^^^^^^^^^^^^^^
   |
   = note: `RecordFieldWrapper` has a derived impl for the trait `Clone`, but this is intentionally ignored during dead code analysis
   = note: `-D dead-code` implied by `-D warnings`
   = help: to override `-D warnings` add `#[allow(dead_code)]`
```

NOTE: playground's nightly is not new enough
[playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=77818b3e2012b708ccb6047b082f66db)

**Comment by apiraino** [comment]
WG-prioritization assigning priority ([Zulip discussion](https://rust-lang.zulipchat.com/#narrow/stream/245100-t-compiler.2Fwg-prioritization.2Falerts/topic/.23120770.20dead_code.20false.20positive.20introduced.20in.20rustc.201.2E78.2E0.E2.80.A6)).

@rustbot label -I-prioritize +P-high

**Comment by apiraino** [comment]
searched nightlies: from nightly-2024-02-01 to nightly-2024-02-08
regressed nightly: nightly-2024-02-08
searched commit range: https://github.com/rust-lang/rust/compare/256b6fb19a2c018eaad4806d2369d1f6a71fc6ec...8ace7ea1f7cbba7b4f031e66c54ca237a0d65de6
regressed commit: https://github.com/rust-lang/rust/commit/d4f6f9ee6a3b24f2cb97b1a5b82963609b93aa33

<details>
<summary>bisected with <a href='https://github.com/rust-lang/cargo-bisect-rustc'>cargo-bisect-rustc</a> v0.6.7</summary>


Host triple: x86_64-unknown-linux-gnu
Reproduce with:
```bash
cargo bisect-rustc --start=2024-02-01 --regress error --script script.sh 
```
</details>

cc @mu001999  @cjgillot

**Comment by Nilstrieb** [comment]
How would you expect the compiler to figure out that it is constructed? Just because there's a value of that type somewhere doesn't mean it is, that would make the lint basically useless.

**Comment by mu001999** [comment]
By the way, the following code does not compile on stable ([play](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=829d96ee099f7b6e4e3a2ec78b59cb74)):

```rust

#![deny(warnings)]

pub struct Bar(u32);

#[repr(transparent)]
struct Wrapper(Bar);

fn try_read<T>(p: &u32) -> T {
    unsafe { std::ptr::read_unaligned(p as *const u32 as *const T) }
}

pub fn foo(p: &u32) -> Bar {
    let Wrapper(bar) = try_read(p);
    bar
}
```

**Comment by Nilstrieb** [comment]
Yeah, this isn't a regression, but a long standing issue (which seems hard to impossible to solve) that's now being exposed a bit more.
Would be nice if we could solve it, but I have doubts about that.

**Comment by oli-obk** [comment]
I don't see a way to solve this without looking through generic function calls (recursively, by looking at the function bodies). That also means it can only be done after monomorphization.

And even then it won't catch e.g. a transmute that produces a None. Instead that would be treated as producing a value of the Some type hypothetically.

Such deeper analyses maybe better suited to external tools that even do partial function instantiation, cross crate analysis...

**Comment by ryanavella** [comment]
At risk of mentioning the obvious, the example provided is not entirely minimal. It will work for any "constructor" of the form `fn foo<T>() -> T`, no trait bounds required.

I'm seeing this in some ffi-heavy code where it didn't trigger previously. In my case the "constructor" looks like this:

```rust
unsafe fn sysctl<T>(mib: &mut [c_int]) -> Option<T> {
    todo!()
}
```

**Comment by shepmaster** [comment]
> It will work for any "constructor" of the form […] no trait bounds required.

Is that not what is shown in https://github.com/rust-lang/rust/issues/120770#issuecomment-1934309081 ?

**Comment by mu001999** [comment]
@rustbot claim

I suddenly realized that this probably didn't need to be done after monomorphization.

We won't meet this warning if specific the type:
```rust
#![deny(warnings)]

pub struct Bar(u32);

#[repr(transparent)]
struct Wrapper(Bar);

fn try_read<T>(p: &u32) -> T {
    unsafe { std::ptr::read_unaligned(p as *const u32 as *const T) }
}

pub fn foo(p: &u32) -> Bar {
    let Wrapper(bar) = try_read::<Wrapper>(p);
    bar
}
```

So I think maybe we can just mark the type live when visiting the corresponding pattern.

I will try it.

**Comment by compiler-errors** [comment]
I believe this is uncovered again since we reverted recent changes to dead code analysis
